### PR TITLE
Adding in correlation metrics (continuing #1782)

### PR DESCRIPTION
**Pull request by mortonjt**
Please complete the following checklist:

* [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/biocore/scikit-bio/blob/master/CONTRIBUTING.md).

* [x] I have documented all public-facing changes in [CHANGELOG.md](https://github.com/biocore/scikit-bio/blob/master/CHANGELOG.md).

* [x] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/biocore/scikit-bio/blob/master/COPYING.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied. **It is your responsibility to disclose code, documentation, or other content derived from external source(s).** If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [ ] **This pull request does not include code, documentation, or other content derived from external source(s).**

**Note:** [REVIEWING.md](https://github.com/biocore/scikit-bio/blob/master/REVIEWING.md) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

This pull request implements the VLR and robust versions of it.  This continues from #1782

### CODE REVIEW

**Comment by mortonjt** [comment]
Hi @wasade it looks like I'm getting some odd checklist errors, thoughts?
```
flake8 skbio setup.py checklist.py
./checklist.py
./checklist.py:215: ChecklistWarning: File skbio/metadata/tests/test_intersection.py has IGNORE directive. Ignoring scikit-bio copyright header validation.
  ChecklistWarning)
Files non-conforming to standard headers as described in
http://scikit-bio.org/docs/latest/development/new_module.html:
    skbio/stats/composition.py

make: *** [Makefile:23: test] Error 1
Error: Process completed with exit code 2.
```

**Comment by wasade** [comment]
It looks like `skbio/stats/composition.py` has an issue with its header, maybe the removed whitespace? I don't know what checklist does -- if you can, recommend running it locally until the module passes

File: skbio/stats/composition.py [hunk 1]
```diff
@@ -807,6 +807,343 @@ def centralize(mat):
...
+    r"""
+    Calculates variance log ratio
+
+    Parameters
+    ----------
+    x : array_like, float
+       a 1-dimensional vector of proportions
+    y : array_like, float
+       a 1-dimensional vector of proportions
+
```
Comment (wasade) [inline skbio/stats/composition.py:822]:
```suggestion
```

File: skbio/stats/composition.py [hunk 2]
```diff
@@ -807,6 +807,343 @@ def centralize(mat):
...
+    # Log transformation
+    x = np.ma.log(x)
+    y = np.ma.log(y)
+
+    # Variance log ratio
+    return np.ma.var(x - y, ddof=ddof)
+
+
+@experimental(as_of="0.5.7")
+def vlr(x: np.ndarray, y: np.ndarray, ddof: int = 1, robust: bool = False):
```
Comment (wasade) [inline skbio/stats/composition.py:897]: here

File: skbio/stats/composition.py [hunk 3]
```diff
@@ -807,6 +807,343 @@ def centralize(mat):
...
+    r"""
+    Calculates variance log ratio
+
+    Parameters
+    ----------
+    x : array_like, float
+       a 1-dimensional vector of proportions
+    y : array_like, float
+       a 1-dimensional vector of proportions
+
```
Comment (wasade) [inline skbio/stats/composition.py:907]: here

File: skbio/stats/composition.py [hunk 4]
```diff
@@ -807,6 +807,343 @@ def centralize(mat):
...
+    ddof: int
+        degrees of freedom
+
```
Comment (wasade) [inline skbio/stats/composition.py:910]: here

File: skbio/stats/composition.py [hunk 5]
```diff
@@ -807,6 +807,343 @@ def centralize(mat):
...
+    float
+         variance log ratio value
+
+    Examples
+    --------
+    No zeros
+    >>> x = [1,2,3]
+    >>> y = [5,8,13]
+    >>> vlr(x,y)
+    >>> 0.01277962183258352
```
Comment (wasade) [inline skbio/stats/composition.py:925]: the user enters this value? i think the `>>>` should not be there, same for the example below

File: skbio/stats/composition.py [hunk 6]
```diff
@@ -807,6 +807,343 @@ def centralize(mat):
...
+
+    # Run backend function
+    if robust:
+        return _robust_vlr(**kwargs)
+    else:
+        return _vlr(**kwargs)
+
+
+@experimental(as_of="0.5.7")
+def _pairwise_vlr(mat: np.ndarray, ddof: int):
```
Comment (wasade) [inline skbio/stats/composition.py:969]: I'll stop noting these at this point :)

**Comment by mortonjt** [comment]
I'm no longer getting that error locally -- let's see what the CI says...

File: skbio/stats/composition.py [hunk 7]
```diff
@@ -807,6 +807,343 @@ def centralize(mat):
...
+
+    ids: array_like, str
+        component names
+
+    ddof: int
+        degrees of freedom
+
+    robust: bool
+        mask zeros at the cost of performance
+
```
Comment (wasade) [inline skbio/stats/composition.py:1088]: validate is not described

**Review by wasade — CHANGES_REQUESTED** [review]
The docstrings need a little work, including extra white space, malformed doctests, and missing the oneliner on the first line. For the functions, we don't currently use annotation in skbio so it would be best to not include right now for consistency.

**Comment by mortonjt** [comment]
Hi @wasade I think this PR is going to need a lot more work - and I don't think I can squeeze this in the near future.  
I recommend we push this in for the next iteration.  That will also buy us more time to look at the numerical issues with the `robust_pairwise_vlr`

**Comment by wasade** [comment]
Sounds good, thanks for confirming

**Comment by mortonjt** [comment]
ok, I'm seeing that some of the runs are passing -- but there are a bunch of errors due to filelock? I guess we'll wait until  #1858 is fixed before merging this in?

**Review by wasade — CHANGES_REQUESTED** [review]
Could a changelog note be added about this? Otherwise +1

**Comment by mortonjt** [comment]
Just added the changelog. Thanks!

File: CHANGELOG.md [hunk 8]
```diff
@@ -14,6 +14,7 @@
 * Added `TreeNode.from_taxdump` for converting taxdump into a tree ([#1810](https://github.com/biocore/scikit-bio/pull/1810)).
 * scikit-learn has been removed as a dependency. This was a fairly heavy-weight dependency that was providing minor functionality to scikit-bio. The critical components have been implemented in scikit-bio directly, and the non-criticial components are listed under "Backward-incompatible changes [experimental]".
 * Python 3.11 is now supported.
+* Adding Variance log ratio estimators in `skbio.stats.composition.vlr` and `skbio.stats.composition.pairwise_vlr` ([#1803](https://github.com/biocore/scikit-bio/pull/1803))
```
Comment (wasade) [inline CHANGELOG.md:17]: Same comment as the other PR, this needs to be under 0.5.8-dev

Comment (mortonjt) [reply CHANGELOG.md]: fixed

**Comment by wasade** [comment]
...conflicts :/

File: skbio/stats/composition.py [hunk 9]
```diff
@@ -107,7 +107,9 @@
 import scipy.stats
 import skbio.util
 from skbio.util._decorator import experimental
+from skbio.stats.distance import DistanceMatrix
 from scipy.sparse import coo_matrix
+>>>>>>> 8983c4e38aa2e581f82f87158561b09a7ee3e5a3
```
Comment (wasade) [inline skbio/stats/composition.py:112]: Merge conflict marks remain

File: skbio/stats/composition.py [hunk 10]
```diff
@@ -802,6 +803,274 @@ def centralize(mat):
     return perturb_inv(mat, cen)
 
 
+<<<<<<< HEAD
```
Comment (wasade) [inline skbio/stats/composition.py:806]: @mortonjt still here. Please do check through the files

Comment (mortonjt) [reply skbio/stats/composition.py]: let's wait for all of the tests to pass first before reviewing.

**Comment by mortonjt** [comment]
Hi @wasade it looks like the tests are now passing :)

**Comment by mortonjt** [comment]
@qiyunzhu @wasade want to merge? :)

**Comment by qiyunzhu** [comment]
@mortonjt Sounds good to me! I don't have the privilege to merge. @wasade you merge? Or grant me permission?

**Review by wasade — APPROVED** [review]

**Pull request merged**